pkg/cfn/manager/tasks_test.go

```diff
@@ -1,104 +1,328 @@
+					Expect(errs[1].Error()).To(Equal("t2.1 always fails"))
+
+					Expect(called11).To(BeTrue())
+					Expect(called21).To(BeTrue())
+					Expect(called31).To(BeTrue())
+					Expect(called32).To(BeTrue())
 				}
-			})
 
-			Context("With an unsuccessful call", func() {
-				JustBeforeEach(func() {
-					errs = sc.RunSingleTask(Task{call, "fail"})
-				})
+				{
+					tasks := &TaskTree{Parallel: false}
+					Expect(tasks.DoAllSync()).To(HaveLen(0))
+				}
+
+				{
+					tasks := &TaskTree{Parallel: false}
+					tasks.Append(&TaskTree{Parallel: false})
+					tasks.Append(&TaskTree{Parallel: true})
+					Expect(tasks.DoAllSync()).To(HaveLen(0))
+				}
+
+				{
+					tasks := &TaskTree{Parallel: true}
+
+					tasks.Append(&taskWithoutParams{
+						info: "t1.1",
+						call: func(errs chan error) error {
+							go func() {
+								time.Sleep(10 * time.Millisecond)
+								errs <- nil
+								close(errs)
+							}()
+							return nil
+						},
+					})
+
+					tasks.Append(&taskWithoutParams{
+						info: "t1.2",
+						call: func(errs chan error) error {
+							go func() {
+								time.Sleep(100 * time.Millisecond)
+								errs <- fmt.Errorf("t1.2 always fails")
+								close(errs)
+							}()
+							return nil
+						},
+					})
+
+					tasks.Append(&taskWithoutParams{
+						info: "t1.3",
+						call: func(errs chan error) error {
+							go func() {
+								errs <- fmt.Errorf("t1.3 always fails")
+								close(errs)
+							}()
+							return nil
+						},
+					})
+
+					tasks.DryRun = true
+
+					Expect(tasks.DoAllSync()).To(HaveLen(0))
+
+					tasks.DryRun = false
+					errs := tasks.DoAllSync()
+					Expect(errs).To(HaveLen(2))
+					Expect(errs[0].Error()).To(Equal("t1.3 always fails"))
+					Expect(errs[1].Error()).To(Equal("t1.2 always fails"))
+				}
+
+				{
+					tasks := &TaskTree{Parallel: false}
+
+					tasks.Append(&taskWithoutParams{
+						info: "t1.1",
+						call: func(errs chan error) error {
+							go func() {
+								time.Sleep(100 * time.Millisecond)
+								errs <- fmt.Errorf("t1.1 always fails")
+								close(errs)
+							}()
+							return nil
+						},
+					})
+
+					tasks.Append(&taskWithoutParams{
+						info: "t1.3",
+						call: func(errs chan error) error {
+							go func() {
+								time.Sleep(150 * time.Millisecond)
+								errs <- nil
+								close(errs)
+							}()
+							return nil
+						},
+					})
+
+					tasks.Append(&taskWithoutParams{
+						info: "t1.3",
+						call: func(errs chan error) error {
+							go func() {
+								errs <- fmt.Errorf("t1.3 always fails")
+								close(errs)
+							}()
+							return nil
+						},
+					})
+
+					tasks.DryRun = true
+
+					Expect(tasks.DoAllSync()).To(HaveLen(0))
+
+					tasks.DryRun = false
+					errs := tasks.DoAllSync()
+					Expect(errs).To(HaveLen(2))
+					Expect(errs[0].Error()).To(Equal("t1.1 always fails"))
+					Expect(errs[1].Error()).To(Equal("t1.3 always fails"))
```

Comment:
Should add another case where a task return an error right way, which should also come on the channel...